delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/model/resources/ResourceRejection.scala

```diff
@@ -0,0 +1,159 @@
+  */
+sealed abstract class ResourceRejection(val reason: String) extends Product with Serializable
+
+object ResourceRejection {
+
+  /**
+    * Rejection returned when a subject intends to retrieve a resource at a specific revision, but the provided revision
+    * does not exist.
+    *
+    * @param provided the provided revision
+    * @param current  the last known revision
+    */
+  final case class RevisionNotFound(provided: Long, current: Long)
+      extends ResourceRejection(s"Revision requested '$provided' not found, last known revision is '$current'.")
+
+  /**
+    * Rejection returned when attempting to create a resource with an id that already exists.
+    *
+    * @param id      the resource identifier
+    */
+  final case class ResourceAlreadyExists(id: Iri) extends ResourceRejection(s"Resource '$id' already exists.")
+
+  /**
+    * Rejection returned when attempting to update a resource with an id that doesn't exist.
+    *
+    * @param id      the resource identifier
+    */
+  final case class ResourceNotFound(id: Iri) extends ResourceRejection(s"Resource '$id' not found.")
+
+  /**
+    * Rejection returned when attempting to create a resource where the passed id does not match the id on the payload.
+    *
+    * @param id        the resource identifier
+    * @param payloadId the resource identifier on the payload
+    */
+  final case class ResourceIdUnexpected(id: Iri, payloadId: Iri)
+      extends ResourceRejection(s"Resource '$id' does not match resource id on payload '$payloadId'.")
+
+  /**
+    * Rejection returned when attempting to create/update a resource where the payload does not satisfy the SHACL schema constrains.
+    *
+    * @param id      the resource identifier
+    * @param schema  the schema for which validation failed
+    * @param report  the SHACL validation failure report
+    */
+  final case class InvalidResource(id: Iri, schema: ResourceRef, report: ValidationReport)
+      extends ResourceRejection(s"Resource '$id' failed to validate against the constrains defined in schema '$schema'")
+
+  /**
+    * Rejection returned when attempting to update/deprecate a resource with a different schema than the resource schema.
+    *
+    * @param id       the resource identifier
+    * @param provided the resource provided schema
+    * @param expected the resource schema
+    */
+  final case class ResourceSchemaUnexpected(id: Iri, provided: ResourceRef, expected: ResourceRef)
+      extends ResourceRejection(
+        s"Resource '$id' is not constrained by the provided schema '$provided', but by the schema '$expected'."
+      )
+
+  /**
+    * Rejection returned when attempting to create a SHACL engine.
+    *
+    * @param id      the resource identifier
+    * @param schema  the resource provided schema
+    * @param details the SHACL engine errors
+    */
+  final case class ResourceShaclEngineRejection(id: Iri, schema: ResourceRef, details: String)
```

Comment:
What are the situations when a simple model would not be accepted by the engine? It seems that the failure is more of a schema problem, not the provided resource.

Reply:
Whatever it is, at this point has to be signaled in the resource STM, since it can happen there. We can change it if needed when schemas are created

Reply:
Please see my comment on porting the validation to the schema (below)